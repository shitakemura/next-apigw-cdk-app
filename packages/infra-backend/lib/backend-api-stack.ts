import * as cdk from "aws-cdk-lib";
import * as lambda_nodejs from "aws-cdk-lib/aws-lambda-nodejs";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as apigatewayv2_alpha from "@aws-cdk/aws-apigatewayv2-alpha";
import * as apigatewayv2_integrations_alpha from "@aws-cdk/aws-apigatewayv2-integrations-alpha";
import { Construct } from "constructs";
import * as path from "path";

export class BackendApiStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // DynamoDB
    const todoTable = new dynamodb.Table(this, "TodoTable", {
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      partitionKey: { name: "userId", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "id", type: dynamodb.AttributeType.STRING },
    });

    // Lambda
    const listTodosLambda = new lambda_nodejs.NodejsFunction(
      this,
      "listTodosHandler",
      {
        entry: path.join(__dirname, "../lambda/listTodos.ts"),
        handler: "handler",
        environment: {
          TODO_TABLE: todoTable.tableName,
        },
      }
    );

    todoTable.grantReadData(listTodosLambda);

    // Http Api
    const todoHttpApi = new apigatewayv2_alpha.HttpApi(this, "TodoHttpApi", {
      corsPreflight: {
        allowHeaders: ["Content-Type", "Authorization"],
        allowMethods: [
          apigatewayv2_alpha.CorsHttpMethod.OPTIONS,
          apigatewayv2_alpha.CorsHttpMethod.GET,
        ],
        allowOrigins: [
          "http://localhost:3000",
          "https://d1ga29t5ymw1db.cloudfront.net",
        ],
      },
    });

    // Http Lambda Integration
    const listTodosIntegration =
      new apigatewayv2_integrations_alpha.HttpLambdaIntegration(
        "listTodosIntegration",
        listTodosLambda
      );

    // Api routes
    todoHttpApi.addRoutes({
      path: "/todos",
      methods: [apigatewayv2_alpha.HttpMethod.GET],
      integration: listTodosIntegration,
    });

    // CrnOutput
    new cdk.CfnOutput(this, "TodoHttpApi Endpoint", {
      value: todoHttpApi.apiEndpoint,
    });
  }
}
