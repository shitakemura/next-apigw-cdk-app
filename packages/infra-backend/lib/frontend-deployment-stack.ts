import * as cdk from "aws-cdk-lib";
import * as cloudFront from "aws-cdk-lib/aws-cloudfront";
import * as s3 from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export class FrontendDeploymentStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const todoBucket = new s3.Bucket(this, "TodoBucket", {
      websiteIndexDocument: "index.html",
      autoDeleteObjects: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    const todoOAI = new cloudFront.OriginAccessIdentity(this, "TodoOAI");
    todoBucket.grantRead(todoOAI);

    const todoWebDestribution = new cloudFront.CloudFrontWebDistribution(
      this,
      "TodoWebDestribution",
      {
        originConfigs: [
          {
            s3OriginSource: {
              s3BucketSource: todoBucket,
              originAccessIdentity: todoOAI,
            },
            behaviors: [{ isDefaultBehavior: true }],
          },
        ],
      }
    );

    new cdk.CfnOutput(this, "TodoWebDestributionName", {
      value: todoWebDestribution.distributionDomainName,
    });
  }
}
