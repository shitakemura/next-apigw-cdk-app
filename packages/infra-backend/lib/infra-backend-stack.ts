import * as cdk from "aws-cdk-lib";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import { Construct } from "constructs";

export class InfraBackendStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const todoTable = new dynamodb.Table(this, "TodoTable", {
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      partitionKey: { name: "id", type: dynamodb.AttributeType.STRING },
    });
  }
}
