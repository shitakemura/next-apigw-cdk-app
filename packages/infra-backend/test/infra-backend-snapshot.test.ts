import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { InfraBackendStack } from "../lib/infra-backend-stack";

test("infra-backend snapshot test", () => {
  const app = new cdk.App();
  const stack = new InfraBackendStack(app, "InfraBackendSnapshotTestStack");
  const template = Template.fromStack(stack);

  expect(template.toJSON()).toMatchSnapshot();
});
