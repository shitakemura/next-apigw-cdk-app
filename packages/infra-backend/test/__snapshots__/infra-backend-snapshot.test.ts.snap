// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Outputs": Object {
    "TodoHttpApiEndpoint": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "TodoHttpApi072A8953",
          "ApiEndpoint",
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TodoHttpApi072A8953": Object {
      "Properties": Object {
        "CorsConfiguration": Object {
          "AllowHeaders": Array [
            "Content-Type",
            "Authorization",
          ],
          "AllowMethods": Array [
            "OPTIONS",
            "GET",
          ],
        },
        "Name": "TodoHttpApi",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "TodoHttpApiDefaultStageEE725B46": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "TodoHttpApi072A8953",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "TodoHttpApiGETtodosE8223B42": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "TodoHttpApi072A8953",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "GET /todos",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "TodoHttpApiGETtodoslistTodosIntegration6E28E076",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "TodoHttpApiGETtodoslistTodosIntegration6E28E076": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "TodoHttpApi072A8953",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "listTodosHandlerA268EC0E",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "TodoHttpApiGETtodoslistTodosIntegrationPermission87920617": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "listTodosHandlerA268EC0E",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "TodoHttpApi072A8953",
              },
              "/*/*/todos",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TodoTable585F1D6B": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "userId",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "userId",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "id",
            "KeyType": "RANGE",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "listTodosHandlerA268EC0E": Object {
      "DependsOn": Array [
        "listTodosHandlerServiceRoleDefaultPolicy196AA5E5",
        "listTodosHandlerServiceRole6D2B3EBF",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "[HASH REMOVED]",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TODO_TABLE": Object {
              "Ref": "TodoTable585F1D6B",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "listTodosHandlerServiceRole6D2B3EBF",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "listTodosHandlerServiceRole6D2B3EBF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "listTodosHandlerServiceRoleDefaultPolicy196AA5E5": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TodoTable585F1D6B",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "listTodosHandlerServiceRoleDefaultPolicy196AA5E5",
        "Roles": Array [
          Object {
            "Ref": "listTodosHandlerServiceRole6D2B3EBF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
